src/Data/Avro/Codec.hs

```diff
@@ -0,0 +1,112 @@
+{-# LANGUAGE BangPatterns        #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE RankNTypes          #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+module Data.Avro.Codec (
+    Codec(..)
+  , nullCodec
+  , deflateCodec
+  ) where
+
+import           Codec.Compression.Zlib.Internal as Zlib
+import qualified Data.Binary.Get                 as G
+import           Data.ByteString                 (ByteString)
+import qualified Data.ByteString                 as BS
+import qualified Data.ByteString.Lazy            as LBS
+
+-- | A `Codec` allows for compression/decompression of a block in an
+-- Avro container according to the Avro spec.
+data Codec = Codec
+  {
+    -- | The name of the codec according to the Avro spec.
+    codecName       :: ByteString
+    -- | Execute a `G.Get` over a chunk of bytes possibly decompressing
+    -- the chunk incrementally.
+    --
+    -- The API is somewhat more complex than say `codecCompress` to allow
+    -- interleaving of decompression and binary decoding while still allowing 
+    -- proper error handling without resorting to async exceptions.
+  , codecDecompress :: forall a. LBS.ByteString -> G.Get a -> G.Get a
+
+    -- | Compresses a lazy stream of bytes.
+  , codecCompress   :: LBS.ByteString -> LBS.ByteString
+  }
+
+-- | `nullCodec` specifies @null@ required by Avro spec.
+-- (see https://avro.apache.org/docs/1.8.1/spec.html#null)
+nullCodec :: Codec
+nullCodec =
+  Codec
+    {
+      codecName = "null"
+    , codecDecompress = \input parser ->
+        case G.runGetOrFail parser input of
+          Right (_, _, x)  -> pure x
+          Left (_, _, err) -> fail err
+    , codecCompress   = id
+    }
+
+-- | `deflateCodec` specifies @deflate@ codec required by Avro spec.
+-- (see https://avro.apache.org/docs/1.8.1/spec.html#deflate)
+deflateCodec :: Codec
+deflateCodec =
+  Codec
+    {
+      codecName       = "deflate"
+    , codecDecompress = deflateDecompress
+    , codecCompress   = deflateCompress
+    }
+
+deflateCompress :: LBS.ByteString -> LBS.ByteString
+deflateCompress =
+  Zlib.compress Zlib.rawFormat Zlib.defaultCompressParams
+
+
+-- | Internal type to help construct a lazy list of
+-- decompressed bytes interleaved with errors if any.
+data Chunk
+  = ChunkRest LBS.ByteString
+  | ChunkBytes ByteString
+  | ChunkError Zlib.DecompressError
+
+
+deflateDecompress :: forall a. LBS.ByteString -> G.Get a -> G.Get a
```

Comment:
Somehow the decoding, deflating, parsing etc is hard for me to follow. Would you mind doing a bit of a play-by-play for `deflateDecompress` and the functions in its `where` clause? The new pattern `ByteString -> Get a -> Get a` comes up several times through the PR, and I don't yet understand it.

Reply:
Ok, let me try to explain more. 

Suppose you have a chunk of bytes `b` that need some post-processing e.g. decompression before you can actually interpret them as meaningful Avro binary. Let's discuss the alternatives.

## decompress :: LBS.ByteString -> LBS.ByteString

```
let 
  decompressed = decompress b
in
  decode decompressed
```

Very simple, very easy, very nice.. Except for async exceptions that might turn up during decompression. This might fail in a fatal way without us being able handle it in case of corrupt data.

## decompress :: LBS.ByteString -> Either Error LBS.ByteString

```
case decompress b of 
  Right decompressed -> 
    decode decompressed
  Left err -> 
    handleError
```

Much better, we are able to handle errors gracefully! But we lost the constant space streaming property! Lazy decompression is a fallacy! To determine if no errors occurred during compression the whole bytestring needs to be decoded first! Means we end up with all the bytes in memory! Bad. Just imagine the sizes of our containers sitting in mem all at once.

Ok, so in order to have the good properties

- graceful error handling without resorting to async exceptions
- and constant time decompression

we need to come up with something with more structure. 

## decompress :: LBS.ByteString -> Get a -> Get a

With incremental decompression we get bytes in chunks as soon as they are available for consumption. Now, we need something that can decode these bytes in an incremental fashion as well. What else is better suited than `Get a`? 

- you can pass your input chunk for chunk
- it signals errors as soon as they happen

so we get the best of both worlds here. Sure, we can simplify the API a bit by saying: 

```
decompress :: LBS.ByteString -> G.Decoder a -> Either String a
```

or similar. But it is the same principle. Or maybe it is the name that is off-putting for you? This is a  weird thing interleaving decompression and decoding after all.

Reply:
I think the interleaved decompression and decoding is some essential complexity that's not easy for me to wrap my head around in the first place. Your explanation helps!